lib/rubocop/cop/discourse/no_add_reference_active_record_migrations.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Discourse
+      # The methods:
+      #
+      # * add_reference
+      # * add_belongs_to
+      # * t.references
+      # * t.belongs_to
+      #
+      # in ActiveRecord migrations are magic, and they all do
+      # some unexpected things in the background. For example, by default
+      # add_reference adds an index at the same time, but not concurrently,
+      # which is a nightmare for large tables.
+      #
+      # Instead, inside a disable_ddl_transaction! migration we should create
+      # the new column (with any defaults and options required) and the index
+      # concurrently using hand-written SQL. This also allows us to handle
+      # IF NOT EXISTS cases, which enable re-runnable migrations. Also we
+      # can pick a better name for the index at the same time.
+      #
+      # @example
+      #
+      # # bad
+      # def up
+      #   add_reference :posts, :image_upload
+      #   # or add_belongs_to :posts, :image_upload
+      #   # or t.references :image_upload when doing create_table do |t|
+      #   # or t.belongs_to :image_upload when doing create_table do |t|
+      # end
+      #
+      # # good
+      # disable_ddl_transaction!
+      # def up
+      #   execute <<~SQL
+      #     ALTER TABLE posts
+      #     ADD COLUMN IF NOT EXISTS image_upload_id bigint
+      #   SQL
+      #
+      #   execute <<~SQL
+      #     CREATE INDEX CONCURRENTLY IF NOT EXISTS
+      #     index_posts_on_image_upload_id ON posts USING btree (image_upload_id)
+      #   SQL
+      # end
+      class NoAddReferenceOrAliasesActiveRecordMigration < Cop
+        MSG = <<~MSG
+          AR methods add_reference, add_belongs_to, t.references, and t.belongs_to are
+          high-risk for large tables and have too many background magic operations.
```

Comment:
I'm not sure if this statement is true because a new column is added with any of the methods so index creation should be fast for a new column.

Reply:
I'm not sure what you mean by this. The issue yesterday as I understand is that we used `add_reference` to an existing table which created a column and added a new index, but the index wasn't added concurrently. It would be fast for a brand new table, true, but in that case why bother banning `t.references`?

Reply:
yea I think we need to clarify this. If we're adding an index on a new column, it shouldn't be breaking anything because the columns are empty.